File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-java/src/main/java/com/example/statemachine/simple/doument/persist/DocumentStateMachine.java
package com.example.statemachine.simple.doument.persist;

import com.example.statemachine.simple.doument.DocumentEvent;
import com.example.statemachine.simple.doument.DocumentState;
import com.example.statemachine.simple.doument.DocumentVo;
import com.example.statemachine.simple.doument.config.DocumentIdChecker;
import com.example.statemachine.simple.doument.config.exception.PersistenceDataIsNotExistException;
import java.util.List;
import java.util.Objects;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.core.RowMapper;
import org.springframework.messaging.support.MessageBuilder;

public class DocumentStateMachine {

  private final JdbcTemplate jdbcTemplate;
  private final DocumentIdChecker documentIdChecker;
  private final DocumentPersistStateMachineHandler handler;

  public DocumentStateMachine(JdbcTemplate jdbcTemplate,
    DocumentPersistStateMachineHandler documentPersistStateMachineHandler) {
    this.jdbcTemplate = jdbcTemplate;
    this.documentIdChecker =  new DocumentIdChecker(jdbcTemplate);
    this.handler = documentPersistStateMachineHandler;
    this.handler.addPersistStateChangeListener(new DocumentStateUpdater(jdbcTemplate));
    this.handler.addPersistStateChangeListener(new DocumentStateChangedHistoryWriter(jdbcTemplate));
  }

  public List<DocumentVo> listDbEntries() {
    return jdbcTemplate.query(
      "SELECT id, state FROM documents", documentRowMapper());
  }

  public void change(final int documentId, final DocumentEvent event)
    throws PersistenceDataIsNotExistException {

    documentIdChecker.isExist(documentId);

    final var d = jdbcTemplate.queryForObject(
      "SELECT id, state FROM documents WHERE id = ?",
      documentRowMapper(), documentId);

    final var message = MessageBuilder.withPayload(event)
      .setHeader("document", documentId)
      .build();

    handler.handleEventWithStateReactively(
        message, Objects.requireNonNull(d).state())
      .block();
  }

  private static RowMapper<DocumentVo> documentRowMapper() {
    return (rs, rowNum) -> new DocumentVo(
      rs.getInt("id"),
      DocumentState.valueOf(rs.getString("state")));
  }

}

File: simple-statemachine-with-persistence/using-statemachine-recipe/simple-statemachine-kotlin/src/main/kotlin/com/example/statemachine/simple/document/persist/DocumentStateMachine.kt
package com.example.statemachine.simple.document.persist

import com.example.statemachine.simple.document.DocumentEvent
import com.example.statemachine.simple.document.DocumentState
import com.example.statemachine.simple.document.DocumentVo
import com.example.statemachine.simple.document.config.DocumentIdChecker
import com.example.statemachine.simple.document.config.exception.PersistenceDataIsNotExistException
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import org.springframework.messaging.Message
import org.springframework.messaging.support.MessageBuilder
import java.sql.ResultSet

class DocumentStateMachine(
  private val jdbcTemplate: JdbcTemplate,
  private val handler: DocumentPersistStateMachineHandler
) {

  private var documentIdChecker: DocumentIdChecker = DocumentIdChecker(jdbcTemplate)

  init {
    handler.addPersistStateChangeListener(DocumentStateUpdater(jdbcTemplate))
    handler.addPersistStateChangeListener(DocumentStateChangedHistoryWriter(jdbcTemplate))
  }

  fun listDbEntries(): List<DocumentVo> {
    return jdbcTemplate.query(
      "SELECT id, state FROM documents", documentRowMapper()
    )
  }

  @Throws(PersistenceDataIsNotExistException::class)
  fun change(documentId: Int, event: DocumentEvent) {

    documentIdChecker.isExist(documentId)

    val d: DocumentVo = jdbcTemplate.queryForObject(
      "SELECT id, state FROM documents WHERE id = ?",
      documentRowMapper(), documentId
    )!!

    val message: Message<DocumentEvent> = MessageBuilder.withPayload<DocumentEvent>(event)
      .setHeader("document", documentId)
      .build()

    handler.handleEventWithStateReactively(message, d.state)
      .block()
  }

  companion object {
    private fun documentRowMapper(): RowMapper<DocumentVo> {
      return RowMapper<DocumentVo> { rs: ResultSet, _: Int ->
        DocumentVo(
          rs.getInt("id"),
          DocumentState.valueOf(rs.getString("state"))
        )
      }
    }
  }
}
